src/snowflake/snowpark/lineage.py

```diff
@@ -0,0 +1,454 @@
+    """
+
+    def __init__(self, session: "snowflake.snowpark.session.Session") -> None:
+        self._session = session
+        self.user_to_system_domain_map = {
+            _UserDomain.FEATURE_VIEW: _SnowflakeDomain.TABLE,
+            _UserDomain.MODEL: _SnowflakeDomain.MODULE,
+        }
+        self.versioned_object_domains = {
+            _UserDomain.FEATURE_VIEW,
+            _UserDomain.MODEL,
+            _SnowflakeDomain.DATASET,
+        }
+
+    def _build_graphql_query(
+        self,
+        object_name: str,
+        object_domain: str,
+        edge_directions: list,
+        object_version: Optional[str] = None,
+    ) -> str:
+        """
+        Constructs a GraphQL query for lineage tracing based on the specified parameters.
+        """
+        properties_string = ", ".join(_ObjectField.GRAPH_ENTITY_PROPERTIES)
+        edge_types_formatted = ", ".join(_EdgeType.list())
+
+        parts = []
+        edge_template = "{direction}: {edge_key}(edgeType:[{edge_types}],direction:{dir}){{{source_key} {{{properties}}}, {target_key} {{{properties}}}}}"
+        for direction in edge_directions:
+            dir_key = (
+                _DGQLFields.OUT
+                if direction == LineageDirection.DOWNSTREAM
+                else _DGQLFields.IN
+            )
+            parts.append(
+                edge_template.format(
+                    edge_key=_DGQLFields.EDGE,
+                    source_key=_DGQLFields.SOURCE,
+                    target_key=_DGQLFields.TARGET,
+                    direction=direction,
+                    dir=dir_key,
+                    edge_types=edge_types_formatted,
+                    properties=properties_string,
+                )
+            )
+
+        object_domain = self.user_to_system_domain_map.get(
+            object_domain.lower(), object_domain
+        )
+
+        parent_param = ""
+        query_object_name = object_name
+        if object_version:
+            parent_param = f', {_ObjectField.PARENT_NAME}:"{object_name}"'
+            query_object_name = object_version
+
+        query_template = '{{{nodeKey}({domainKey}: {domain}, {nameKey}:"{name}"{parent_param}) {{{edges}}}}}'
+        query = query_template.format(
+            nodeKey=_DGQLFields.NODE,
+            domainKey=_ObjectField.DOMAIN,
+            nameKey=_ObjectField.NAME,
+            domain=object_domain.upper(),
+            name=query_object_name,
+            parent_param=parent_param,
+            edges="".join(parts),
+        )
+
+        return f"select SYSTEM$DGQL('{query}')"
+
+    def _get_lineage(
+        self,
+        object_name: str,
+        object_domain: str,
+        directions: list,
+        object_version: Optional[str] = None,
+        current_depth=1,
+    ) -> List[Tuple[VariantType, VariantType, StringType, int]]:
+        """
+        Constructs and executes a query to trace the lineage of a given entity at a single depth level.
```

Comment:
maybe mention the output schema in comment: source, target, dir, depth. So it's earlier to interpret later